### PR TITLE
perf(meta): reduce memory usage which are used to store hummock version

### PR DESCRIPTION
**Pull request by soundOfDestiny**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

reduce memory usage which are used to store hummock version
as discussed in #3689 
> Currently we keep both the `HummockVersionDelta` and the full `HummockVersion` in memory for all pinned version id, which looks redundant to me. I think we only need to keep three things in memory:
> 
> 1. The persisted full `HummockVersion`
> 2. The current `HummockVersion` (optional, it can be re-constructed by 1) and 3))
> 3. `HummockVersionDelta` between 1) and 2).
> 
> ```
> | HummockVersion           |  -> persisted, in-memory
> | VersionDelta1            |  -> persisted, in-memory
> | VersionDelta2            |  -> persisted, in-memory
>       ...
> | VersionDeltaX            |  -> persisted, in-memory
> | HummockVersion (current) |  -> in-memory
> ```

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/meta/src/hummock/hummock_manager.rs [hunk 1]
```diff
@@ -178,7 +178,8 @@ macro_rules! start_measure_real_process_timer {
 #[derive(Default)]
 struct Versioning {
     current_version_id: CurrentHummockVersionId,
-    hummock_versions: BTreeMap<HummockVersionId, HummockVersion>,
+    current_version: Option<HummockVersion>,
```
Comment (Little-Wallace) [inline src/meta/src/hummock/hummock_manager.rs:181]: Why it need to be `Option`? 
Could we have an initial version?

Comment (soundOfDestiny) [reply src/meta/src/hummock/hummock_manager.rs:181]: > Why it need to be `Option`?
> Could we have an initial version?

initial version is generated when `load_mem_store` is called, which is too late

Comment (soundOfDestiny) [reply src/meta/src/hummock/hummock_manager.rs]: > Why it need to be `Option`? Could we have an initial version?

fixed

File: src/meta/src/hummock/hummock_manager.rs [hunk 2]
```diff
@@ -178,7 +178,8 @@ macro_rules! start_measure_real_process_timer {
 #[derive(Default)]
 struct Versioning {
     current_version_id: CurrentHummockVersionId,
-    hummock_versions: BTreeMap<HummockVersionId, HummockVersion>,
+    current_version: HummockVersion,
+    hummock_versions: BTreeMap<HummockVersionId, Vec<u64>>,
```
Comment (hzxa21) [inline src/meta/src/hummock/hummock_manager.rs:182]: `Vec<u64>` -> `Vec<HummockSSTableId>` to make the code more readable. Can you also add a doc to explain:
1. what versions are kept in this map?
2. why we need to track SST ids for each version?

**Review by hzxa21 — COMMENTED** [review]
Now that we have incremental version implemented in meta. Can we re-run the hummock service benchmark again to see how much improvement we can get?

File: src/meta/src/hummock/hummock_manager.rs (hunk 2 above)
Comment (soundOfDestiny) [reply src/meta/src/hummock/hummock_manager.rs:182]: > `Vec<u64>` -> `Vec<HummockSSTableId>` to make the code more readable. Can you also add a doc to explain:
> 
> 1. what versions are kept in this map?

all

> 2. why we need to track SST ids for each version?

given pin counts of each version, respectively, we need to determine whether a SST can be deleted in vacuum process

**Comment by zwang28** [comment]
> Now that we have incremental version implemented in meta. Can we re-run the hummock service benchmark again to see how much improvement we can get?

I think we may do it after https://github.com/singularity-data/risingwave/issues/4037 , where stale_sstables and sstable_id_infos are further removed.

File: src/meta/src/hummock/hummock_manager.rs (hunk 2 above)
Comment (hzxa21) [reply src/meta/src/hummock/hummock_manager.rs]: > > `Vec<u64>` -> `Vec<HummockSSTableId>` to make the code more readable. Can you also add a doc to explain:
> > 
> > 1. what versions are kept in this map?
> 
> all
> 
> > 2. why we need to track SST ids for each version?
> 
> given pin counts of each version, respectively, we need to determine whether a SST can be deleted in vacuum process

 I know. I mean you can add documentation in the codes :)

Comment (soundOfDestiny) [reply src/meta/src/hummock/hummock_manager.rs]: > we need to determine whether a SST can be deleted in vacuum process

fixed. just added some

**Review by hzxa21 — APPROVED** [review]

**Pull request merged**